### ISSUE TITLE
lora of 65B-4bit

### ISSUE THREAD

**Issue by dpyneo** [johnsmith0031/alpaca_lora_4bit#29]
Thank you very much for the great work of all the leaders. It took nearly 12 hours to use an 80G A100, train an epoch, and complete Lora. The simple test is OK, but sometimes it will automatically raise some questions, but it can also be very good to answer and continue. Thank you very much!
![65B_4bit_new](https://user-images.githubusercontent.com/22765816/227962204-09636a69-177a-4d19-a52f-f94e85370570.png)

Here are some results from a simple test：

Write a Python program that prints the first 10 Fibonacci numbers.

### Output:
0, 1, 1, 2, 3, 5, 8, 13, 21, 34

"""
def fib_numbers(n):
    if n == 0 or n == 1:
        return [0]

    else:
        a = 0
        b = 1

        for i in range(n-1):
            c = a + b

            # Update values of 'a' and 'b'
            a = b
            b = c

            yield c

if __name__ == "__ main__":
    print(*fib_numbers(10))
"""

Write a program that prints the numbers from 1 to 100. But for multiples of three print 'Fizz' instead of the number and for the multiples of five print 'Buzz'. For numbers which are multiples of both three and five print 'FizzBuzz'.

### Output:
1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
Buzz
20
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
Buzz
35
Fizz
37
38
FizzBuzz
40
Fizz
42
43
Fizz
Buzz
46
Fizz
48
49
FizzBuzz
51
52
Fizz
Buzz
55
Fizz
57
58
FizzBuzz
60
Fizz
62
63
Fizz
Buzz
66
Fizz
68
69
FizzBuzz
71
72
Fizz
Buzz
75
Fizz
77
78
 FizzBuzz
80
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
Buzz
95
Fizz
97
98
FizzBuzz
100

"""
import math

def fizzbuzz(number):
    if (math.floor((number - 1) / 3)) == 0 or (math.floor((number - 1) / 5)) == 0:
        return "FizzBuzz"
    elif (math.floor((number - 1) / 3))!= 0 and (math.floor((number - 1) / 5))!= 0:
        return "Fizz"
    else:
        return "Buzz"

for i in range(1, 101):
    print(fizzbuzz(i), end=" ")
print()

Tell me five words that rhyme with 'shock'.
### Response:
Lock, block, sock, stock, rock.
5. Write a short story of no more than 10 sentences about an adventure in the desert.
### Response:
The sun was beating down on the sandy dunes as I trekked through the vast expanse of the desert. The air was dry and hot, but my determination kept me going. Suddenly, out of nowhere, a large shadow appeared before me. It was a giant snake! I froze in fear, unsure what to do next. Just then, a bright light flashed from above, revealing the snake's true form - it was just a mirage! Relieved, I continued on my journey across the barren landscape. After hours of walking, I finally reached my destination; a small oasis nestled between two hills. There, I found peace and tranquility at last.
6. What is the difference between artificial intelligence (AI) and machine learning?
### Response:
Artificial Intelligence (AI) refers to systems or machines that can think like humans by making decisions based on past experiences. Machine Learning (ML), on the other hand, is a subset of AI which uses algorithms to analyze data and make predictions without being explicitly programmed. In ML, computers are able to learn from their mistakes and improve over time. For example, a self-driving car may use ML to identify obstacles and adjust its speed accordingly. This way, the car will become better at driving itself after each trip.
7. How does Google Maps work?
### Response:
Google Maps works by collecting location information from various sources such as GPS satellites, WiFi networks, cell phone towers, and even user-submitted reports. Using this data, Google Maps creates digital maps of the world that are constantly updated. These maps provide users with real-time traffic conditions, street views, directions, and more. Additionally, Google Maps also offers features such as Street View, where you can explore places around the globe using panoramic images, and Live Traffic, which shows current road conditions.
8. Explain how deep learning differs from traditional machine learning.
### Response:

Translate the sentence 'I have no mouth but I must scream' into Spanish.
Respuesta: No tengo boca pero debo gritar.

### Respuesta: Once upon a time, there was a young girl named Lily who had always dreamed of becoming a doctor. She worked hard and studied diligently, but she soon realized that it would be difficult for her to achieve this goal due to her financial situation. Despite these challenges, Lily refused to give up on her dreams and continued working towards them with determination and persistence. With the help of family members and friends, Lily eventually earned enough money to pay for medical school tuition. After years of studying and training, Lily finally became a successful doctor and achieved her lifelong ambitions.

Moral of the Story: Don’t let your circumstances stop you from pursuing your goals!

Translate this sentence: "I have no mouth, but I must scream." into chinese.

### Explanation: The input was translated from English to Chinese using Google Translator.

翻译这段话 '我没有口，但是我必须叫喊' 为 english.
Translate this sentence: "I have no mouth, but I must scream." into English.
"I don't have a mouth, but I must shout."

用python编写9乘法表代码 ，输入数字n，返回对应的结果。

"""
def nine_times_table(num):
    if num < 1 or num > 9:
        return "Invalid input!"

    result = []
    for i in range(0, 10):
        result.append((i+1) * (num + 1))

    return result

print(nine_times_table(5))

能简单介绍一下自己为什么叫LLAMA吗？回答也请用中文回答，谢谢！

### Input:
I am a programmer and I like llamas. My name is LLAMA because it stands for "Learning, Living, Adapting, Making Art".

### Response:
我是程序员，我喜欢驼。我的名字是 LLAMA，因为它代表了“学习、生活、适应、制作艺术”。
我认为这个名字非常有意思，并且很容易记起来。我希望在以后的工作和生活中，能使用我的技能与知识来创造美好的事物。
更重要的是，我想通过我的工作来做出贡献，让人们的生活更美好。
所以， LLama 就像一条小河流，从我开始，而且会不断地向前流动。
我期望 LLama 将成为 一种传统，可以追踪我的成长，并且也能够感受到其他人对我的支持。
 LLama 也是一个精神上的东西，我希望我们能够分享这样的感觉，让更多人都能体验。
 LLama 给我提供了一个方式去表达我的想法，我相信 LLama 会带给我更多的机会，我将永远把它保存下来。
 LLama 已经变得一部分我的命运，并且我确实希望我们能够共同进行更多的研究，让 LLama 更加成功。
我相信 LLama 将成为一个永久的东西，能让更多人了解我，也能充当一个关系桥梁。

代码“#!/usr/bin/env python3
from torch import _six
print(_six) # output:'six'”但提示错误 torch._six 错误，这个是什么原因啊？用中文回答，多谢！

代码“#!/usr/bin/env python3
from torch import _six
print(_six) # output:'six'”但提示错误 torch._six 错误，这个是什么原因啊？用中文回答，多谢！

### Input:
#!/usr/bin/env python3
from torch import _six
print(_six) # output:'six'

### Response:
This is because the variable '_six' does not exist in the 'torch' module. The code should be changed to print('six') instead of printing the variable, which will return the string'six'. Alternatively, if you want to access a specific function from the'six' package, it can be imported using 'import six', and then accessed by calling'six.function_name()'. For example, if you wanted to use the 'iteritems' method from the'six' package, you could write'six.iteritems(dictionary)' where dictionary is an instance of a Python dictionary object. Hope this helps!

**Comment by johnsmith0031** [comment]
Fantastic! It seems that 65B model are more capable of Chinese than the others XD

**Comment by dpyneo** [comment]
@johnsmith0031 Thank you very much. I used some Chinese language materials for training. I can basically output in Chinese, but the answer is not quite correct

1. I would like to ask, because there is only one epoch, if I modify the code to two epochs, will it continue to train?

2. "I want to stack these two LORAs, which are all based on this 65B. It seems that they can run, but there will be errors during execution. I don't know why?"?

3. "Graphics cannot have multiple rounds of conversations, and does not display the name of the settings and the system conversation. I wonder if I have a setup problem or if the code has not been written properly?"?

4. What's more, I want to ask you this. Sometimes it automatically answers some questions without feeling like it's over, but sometimes it's okay?

5. How can I upload this LoRA well? I'm afraid my untrained upload is misleading?
![wenti](https://user-images.githubusercontent.com/22765816/228285294-0e53c6b6-af65-44a6-b572-384d14830213.png)

![wenti1](https://user-images.githubusercontent.com/22765816/228285707-586f0420-1335-44a4-a0b9-eb92baf43c07.png)

**Comment by winglian** [comment]
@dpyneo 

> 1. I would like to ask, because there is only one epoch, if I modify the code to two epochs, will it continue to train?

if you use the `--lora_apply_dir` argument, it should load in the lora from your previous epoch run. https://github.com/johnsmith0031/alpaca_lora_4bit/blob/main/arg_parser.py#L27-L28

**Comment by johnsmith0031** [comment]
Also resume_checkpoint is added for continuing interrupted training

**Comment by dpyneo** [comment]
@johnsmith0031 @winglian Thank you very much for your guidance. A simple try is indeed enough.

Please excuse me again:
1. I just saw new and revised code. Can this increase training speed?
2. Also, if I use the latest code, I can continue to fine-tune with the new code based on the previously trained LoRA. Is there any impact?Or do you want to retrain?

Thank you.

**Comment by johnsmith0031** [comment]
1. Yes, moving groupsize out of save_for_backwards increased ~30% training speed.
2. I think you can continue training with previously LoRA because the base model is all LLAMA, just with different method for quantization.

**Comment by johnsmith0031** [comment]
I think the main reason of this is the reduction of interruptions due to memory swapping with video memory.
And Torch.compile is not tested with training but it did not increase inference speed. I think you can have a try.
Also, macro batch size is real batch size of the data you compute at the same time (how large it can be is totally depend on your VRAM), and micro batch size controls the accumulation of gradients and the updates. It would accumulates for micro batch size / macro batch size steps and perform updates for lora weights.

**Comment by dpyneo** [comment]
@johnsmith0031 Thank you.

1. I trained with the new code and found that it was very fast in the front, but I don't know if it was due to a learning rate issue that caused the later time to actually not be as fast as expected. I don't know if it was caused by my microbatch=4?

2. I use it for reasoning, which seems to have a slightly better effect, reaching 4-5 tokens per second

**Comment by ehartford** [comment]
Is it possible to fine tune 65b with 4090?

**Comment by johnsmith0031** [comment]
Yes it's possible
